integration/fuse/src/main/java/alluxio/fuse/file/MemoryCacheFileInStream.java

```diff
@@ -0,0 +1,236 @@
+   * @throws AlluxioException
+   */
+  public MemoryCacheFileInStream(URIStatus uriStatus, FileInStream fileInStream)
+      throws IOException, AlluxioException {
+    mStatus = uriStatus;
+    mFileInStream = fileInStream;
+  }
+
+  @Override
+  public long remaining() {
+    return mStatus.getLength() - mPos;
+  }
+
+  @Override
+  public int positionedRead(long position, byte[] buffer, int offset, int length)
+      throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return mPos;
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    mPos = pos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    int length = Math.min(len, b.length - off);
+    int read = 0;
+    while (length > 0) {
+      byte[] bytes = getCacheBytes(length);
+      if (bytes.length == 0) {
+        break;
+      }
+      int copyLen = Math.min(length, bytes.length);
+      System.arraycopy(bytes, 0, b, off + read, copyLen);
+      mPos += copyLen;
+      read += copyLen;
+      length -= copyLen;
+    }
+    return read == 0 ? -1 : read;
+  }
+
+  @Override
+  public int read(ByteBuffer buf) throws IOException {
+    return super.read(buf);
+  }
+
+  @Override
+  public int read(ByteBuffer byteBuffer, int off, int len) throws IOException {
+    byteBuffer.position(off);
+    int length = Math.min(len, byteBuffer.capacity() - off);
+    byte[] data = new byte[length];
+    int read = read(data);
+    if (read > 0) {
+      byteBuffer.put(data, 0, read);
+    }
+    return read;
+  }
+
+  @Override
+  public void unbuffer() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int available() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void close() throws IOException {
+    mFileInStream.close();
+  }
+
+  @Override
+  public synchronized void mark(int readlimit) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public synchronized void reset() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean markSupported() {
+    throw new UnsupportedOperationException();
+  }
+
+  private byte[] getCacheBytes(int length) throws IOException {
+    long posKey = (mPos / CACHE_SIZE) * CACHE_SIZE;
+    int start = (int) (mPos % CACHE_SIZE);
+    try {
+      if (mStatus.getLength() <= mPos) {
+        return new byte[0];
+      }
+      byte[] bytes = CACHE.get(
+          new CacheKey(mStatus.getPath(), mStatus.getLastModificationTimeMs(), posKey),
```

Comment:
can u double check on this lastmodificationtime? does openFile call getStatus which will set the lastmodificationtime on READ access ?

Reply:
Since I am using externally provided status and fileInStream, these two should be consistent when passed in. Additionally, if the file has been modified, it should be the responsibility of the fileInStream to report an error during reading, rather than checking it here.

Reply:
no i was meaning to say that mStatus.lastmodificationtime gets changed when making getStatus call (getFileInfo call on master), but it was actually lastAccessTime, so never mind on this comment :)